custom_tracing/src/lib.rs

```diff
@@ -0,0 +1,130 @@
+use anyhow::Result;
+use futures::{channel::mpsc::channel, SinkExt, StreamExt};
+use notify::{event::DataChange, Config, RecommendedWatcher, RecursiveMode, Watcher};
+use std::{fs, path::Path};
+use tracing::Span;
+use tracing_subscriber::{
+    layer::SubscriberExt,
+    reload::{self, Handle},
+    util::SubscriberInitExt,
+    EnvFilter, Registry,
+};
+
+#[cfg(feature = "grpc")]
+pub mod grpc_layer;
+#[cfg(feature = "http-1")]
+pub mod http_layer;
+
+pub async fn init(og_filter: String, file: String) -> Result<()> {
+    let (filtered_layer, reload_handle) =
+        reload::Layer::new(tracing_subscriber::EnvFilter::new(og_filter.clone()));
+
+    tracing_subscriber::registry()
+        .with(filtered_layer)
+        .with(tracing_subscriber::fmt::layer())
+        .init();
+
+    let state = State {
+        og_filter: og_filter.clone(),
+        reload_handle,
+    };
+
+    tokio::spawn(async move {
+        if let Err(e) = watch_file(state, file).await {
+            tracing::warn!("error: {:?}", e)
+        }
+    });
+
+    Ok(())
+}
+
+pub fn record<T>(field: &str, value: T)
+where
+    T: ToString,
+{
+    Span::current().record(field, &tracing::field::display(value.to_string()));
+}
+
+#[derive(Clone)]
+pub struct State {
+    pub og_filter: String,
+    pub reload_handle: Handle<EnvFilter, Registry>,
+}
+
+impl State {
+    fn handle_change(&self, content: String) -> Result<()> {
+        if content.is_empty() {
+            self.handle_delete()
+        } else {
+            match tracing_subscriber::EnvFilter::try_new(content.clone()) {
+                Err(_e) => tracing::warn!(filter = content, "failed to parse filter {:?}", _e),
+                Ok(new_filter) => {
+                    self.reload_handle.modify(|filter| *filter = new_filter)?;
+                    tracing::info!(filter = content, "updated");
+                }
+            }
+            Ok(())
+        }
+    }
+
+    fn handle_delete(&self) -> Result<()> {
+        let new_filter = self.og_filter.clone();
+
+        self.reload_handle
+            .modify(|filter| *filter = tracing_subscriber::EnvFilter::new(new_filter.clone()))?;
+
+        tracing::info!(filter = new_filter, "deleted tracing file, updated");
+        Ok(())
+    }
+}
+
+async fn watch_file(state: State, file: String) -> Result<()> {
+    let (mut tx, mut rx) = channel(1);
```

Comment:
It doesn't really look like this function needs to be async? You can just move the match statement into the watcher closure, then call this function in a spawn blocking

Reply:
done